### PR TITLE
Expanding Test Configurations & Prioritised Experience Replay

### PR DESCRIPTION
**Pull request by Matteo299**
# Description

- Expanded test configurations by testing graphs with every node as a source.
- Implemented and enabled Prioritized Experience Replay for training l-dgn, hl-dgn and dgn-r.

## Type of change

Changes to previous code:
- In core.py, testing episodes now consider a different source every time. Also, random movement function is now seeded as every other random function in the framework.
- Some minor changes in hl_dgn.py, l_dgn.py and dgn_r.py to increase the number of testing episodes and to enable tianshou's Prioritized Replay Buffer.
- Changes to dgn.py to enable Prioritized Buffer Replay mechanism for dgn-r.

### CODE REVIEW

File: dgn_r.py [hunk 1]
```diff
@@ -228,7 +228,7 @@ def watch(
 
     collector = MultiAgentCollector(masp_policy, env, exploration_noise=False,
                                     number_of_agents=args.n_agents)
-    result = collector.collect(n_episode=args.test_num)
+    result = collector.collect(n_episode=args.test_num * args.n_agents)
```
Comment (Matteo299) [inline dgn_r.py:231]: Number of testing episodes are now: # of testing graphs *  # of agents

File: dgn_r.py [hunk 2]
```diff
@@ -330,7 +330,7 @@ def reward_metric(rews):
             max_epoch=args.epoch,
             step_per_epoch=args.step_per_epoch,
             step_per_collect=args.step_per_collect,
-            episode_per_test=args.test_num,
+            episode_per_test=args.test_num * args.n_agents,
```
Comment (Matteo299) [inline dgn_r.py:333]: Same as above

File: graph_env/env/utils/core.py [hunk 3]
```diff
@@ -193,6 +186,7 @@ def __init__(
             self.graphs = cycle(glob.glob(f"graph_topologies/testing_{self.num_agents}/*"))
         else:
             self.graphs = glob.glob(f"graph_topologies/training_{self.num_agents}/*")
+        self.tested_agent = 0
```
Comment (Matteo299) [inline graph_env/env/utils/core.py:189]: New instance variable to track the current node which needs to been selected as a source

File: graph_env/env/utils/core.py [hunk 4]
```diff
@@ -306,7 +300,7 @@ def update_position(self, step):
         pos = nx.get_node_attributes(self.graph, "pos")
 
         # Given the step size, compute the x and y movement for each agent
-        offset_x, offset_y = compute_random_movement(self.num_agents, step)
+        offset_x, offset_y = self.compute_random_movement(step)
```
Comment (Matteo299) [inline graph_env/env/utils/core.py:303]: compute_random_movement is not static anymore

File: graph_env/env/utils/core.py [hunk 5]
```diff
@@ -350,12 +344,19 @@ def update_two_hop_neighbors(self, agent):
         # Agent can't be two hop neighbor of himself
         agent.two_hop_neighbours_ids[agent.id] = 0
 
+    # Method that calculate random movement for the agents if the graph is dynamic
+    def compute_random_movement(self, step):
+        ox = [step * self.np_random.uniform(-1, 1) for _ in range(self.num_agents)]
```
Comment (Matteo299) [inline graph_env/env/utils/core.py:349]: np_random seeded is now used

File: graph_env/env/utils/core.py [hunk 6]
```diff
@@ -350,12 +344,19 @@ def update_two_hop_neighbors(self, agent):
...
+        return ox, oy
+
     def reset(self):
         if self.random_graph:
             self.graph = create_connected_graph(n=self.num_agents, radius=self.radius)
         elif not self.is_graph_fixed:
             if self.is_testing:
-                self.graph = load_graph(next(self.graphs))
+                if self.tested_agent == 0:
+                    self.graph = load_graph(next(self.graphs))
```
Comment (Matteo299) [inline graph_env/env/utils/core.py:359]: A new graph is loaded only if the new source is node 0

File: graph_env/env/utils/core.py [hunk 7]
```diff
@@ -367,9 +368,14 @@ def reset(self):
                        for i in range(self.num_agents)]
         # Includes origin message
         self.messages_transmitted = 0
-        random_agent = self.agents[0] if self.is_testing else self.np_random.choice(self.agents)
+        random_agent = self.agents[self.tested_agent] if self.is_testing else self.np_random.choice(self.agents)
         self.origin_agent = random_agent.id
 
+        if not self.is_graph_fixed and self.is_testing:
+            self.tested_agent += 1
```
Comment (Matteo299) [inline graph_env/env/utils/core.py:375]: Test the same graph with the following node as a source

File: graph_env/env/utils/core.py [hunk 8]
```diff
@@ -367,9 +368,14 @@ def reset(self):
...
+            if self.tested_agent == self.num_agents:
+                self.tested_agent = 0
```
Comment (Matteo299) [inline graph_env/env/utils/core.py:377]: If we tested all the source, load the next graph

File: dgn_r.py [hunk 9]
```diff
@@ -119,6 +119,9 @@ def get_parser() -> argparse.ArgumentParser:
     parser.add_argument("--n-warmup-steps", type=int, default=3)
     parser.add_argument("--timeout", type=float, default=None)
 
+    parser.add_argument("--alpha", type=float, default=0.6)
+    parser.add_argument("--beta", type=float, default=0.4)
```
Comment (Matteo299) [inline dgn_r.py:123]: Two new args parameters are added, _alpha_ determines how much prioritization is used.
_Beta_ instead influences the importance-sampling (IS) mechanism to correct the prioritization bias.

File: dgn_r.py [hunk 10]
```diff
@@ -263,9 +266,11 @@ def train_agent(
     train_collector = MultiAgentCollector(
         masp_policy,
         train_envs,
-        VectorReplayBuffer(args.buffer_size,
-                           len(train_envs) * len(agents),
-                           ignore_obs_next=True),
+        PrioritizedVectorReplayBuffer(args.buffer_size,
```
Comment (Matteo299) [inline dgn_r.py:269]: Tianshou's PrioritizedVectorReplayBuffer used only for training, not for testing.
_Alpha_ and _beta_ are mandatory parameters.

File: graph_env/env/utils/policies/dgn.py [hunk 11]
```diff
@@ -18,7 +18,9 @@ def learn(self, batch: Batch, **kwargs: Any) -> Dict[str, float]:
             self.sync_weight()
         self.optim.zero_grad()
         weight = batch.pop("weight", 1.0)
-        partial_loss = torch.zeros((len(batch), ), device='cuda')
+        device = 'cuda' if torch.cuda.is_available() else None
+        partial_loss = torch.zeros((len(batch), ), device=device)
+        td_errors = torch.zeros((len(batch), ), device=device)
```
Comment (Matteo299) [inline graph_env/env/utils/policies/dgn.py:23]: Cuda is enabled only if drivers are available. This allows the code to work seamlessly with GPUs not produced by Nvidia.

File: graph_env/env/utils/policies/dgn.py [hunk 12]
```diff
@@ -31,11 +33,12 @@ def learn(self, batch: Batch, **kwargs: Any) -> Dict[str, float]:
             q = q[np.arange(len(q)), neighbour_obs.act]
 
             returns = to_torch_as(neighbour_obs.rew.flatten(), q)
-            partial_loss[i] = ((returns - q).pow(2) * weight).mean()
+            partial_loss[i] = ((returns - q).pow(2)).mean()
+            td_errors[i] = (returns - q).mean()
```
Comment (Matteo299) [inline graph_env/env/utils/policies/dgn.py:37]: td_errors are calculated here

File: graph_env/env/utils/policies/dgn.py [hunk 13]
```diff
@@ -31,11 +33,12 @@ def learn(self, batch: Batch, **kwargs: Any) -> Dict[str, float]:
...
 
-        loss = partial_loss.mean()
+        loss = (partial_loss * weight).mean()
```
Comment (Matteo299) [inline graph_env/env/utils/policies/dgn.py:39]: The product with the weights is moved here

File: graph_env/env/utils/policies/dgn.py [hunk 14]
```diff
@@ -31,11 +33,12 @@ def learn(self, batch: Batch, **kwargs: Any) -> Dict[str, float]:
...
 
-        batch.weight = 1 # prio-buffer TODO: not handling prio-buffer yet
+        batch.weight = td_errors # prio-buffer
```
Comment (Matteo299) [inline graph_env/env/utils/policies/dgn.py:41]: This mechanism allows the prioritized replay buffer to work

File: dgn_r.py [hunk 15]
```diff
@@ -99,6 +99,13 @@ def get_parser() -> argparse.ArgumentParser:
         help="Enable dynamic graphs"
     )
 
+    parser.add_argument(
+        "--prio-buffer",
+        default=False,
+        action="store_true",
+        help="Enable prioritized experience replay"
+    )
```
Comment (Matteo299) [inline dgn_r.py:107]: Flag to enable prioritized experience replay

File: dgn_r.py [hunk 16]
```diff
@@ -259,13 +269,20 @@ def train_agent(
...
                                             optim=optim)
 
+    train_replay_buffer = PrioritizedVectorReplayBuffer(args.buffer_size,
+                                                        len(train_envs) * len(agents),
+                                                        ignore_obs_next=True,
+                                                        alpha=args.alpha,
+                                                        beta=args.beta) if args.prio_buffer else \
+                          VectorReplayBuffer(args.buffer_size,
+                                             len(train_envs) * len(agents),
+                                             ignore_obs_next=True)
```
Comment (Matteo299) [inline dgn_r.py:279]: Here the program decides which replay buffer to use according to the flag

**Review by RaffaeleGalliera — APPROVED** [review]
Looks good! :+1:

**Pull request merged**